mount-control: step 2

mardy opened the pull request:
This is the second part of the mount-control saga: this adds the interface itself and a spread test.

The full (draft) MR is [here](https://github.com/snapcore/snapd/pull/10473), whereas the first part (which this branch depends on) is https://github.com/snapcore/snapd/pull/10653.

pedronis commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,310 @@
+	if ok {
+		interfaceList, ok := value.([]interface{})
+		if !ok {
+			return nil, fmt.Errorf(`mount-control "mount::%s" must be an array of strings (got %q)`, fieldName, value)
```
we don't use "::" as separator like this,  we should probably either use "." or be vervose and use "mount entry "%s"

pedronis replied:
all the validation error cases need test coverage

mardy replied:
Done.

pedronis commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,310 @@
+
+		what, ok := mount["what"].(string)
+		if !ok {
+			return fmt.Errorf(`mount-control "mount::what" must be a string`)
```
same about "::" and everywhere

mardy replied:
Done.

pedronis commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,310 @@
+func (iface *mountControlInterface) BeforePreparePlug(plug *snap.PlugInfo) error {
+	// The systemd.ListMountUnits() method works by issuing the command
+	// "systemctl show *.mount", but globbing was only added in systemd v209.
+	if err := systemd.EnsureAtLeast(209); err != nil {
```
I think we might want to check for this only on connection and not simply plug existence, so that would be BeforeConnectPlug. Maybe @anonymouse64 has also an opinion on this

pedronis reviewed (commented):
did a first pass, some comments

mardy commented on interfaces/utils/path_patterns.go:
```diff
@@ -0,0 +1,154 @@
+	return pp, nil
+}
+
+func (pp *PathPattern) Matches(path string) bool {
```
Done.

mardy commented on interfaces/utils/path_patterns.go:
```diff
@@ -0,0 +1,154 @@
+	return regex, nil
+}
+
+func NewPathPattern(pattern string) (*PathPattern, error) {
```
Done.

mardy commented on tests/main/interfaces-mount-control/task.yaml:
```diff
@@ -0,0 +1,42 @@
+
+    echo "Verify that the snap can perform a mount"
+    mkdir -p "$MOUNT_DEST"
+    "${SNAP_NAME}".cmd mount -o bind,rw "$MOUNT_SRC" "$MOUNT_DEST"
```
Done.

alexmurray commented on interfaces/utils/path_patterns.go:
```diff
@@ -0,0 +1,154 @@
+// createRegex converts the apparmor-like glob sequence into a regex. Loosely
+// using this as reference:
+// https://gitlab.com/apparmor/apparmor/-/blob/master/parser/parser_regex.c#L107
+func createRegex(pattern string) (string, error) {
```
It would be great to get @jrjohansen  to give this function a once-over.

jrjohansen replied:
how faithful does the conversion from aa globbing to re need to be? The conversion does not appear to be correct in some cases

alexmurray commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,320 @@
+	"dirsync",
+	"iversion",
+	"lazytime",
+	"mand",
```
I don't think we should allow `mand` - according to the `fcntl(2)` man page, the Linux implementation of mandatory locking is buggy so I fear this could lead to a state where we are unable to unmount a file-system. So I think we should not allow this unless there is a strong use-case for it.

mardy replied:
Removed.

alexmurray commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,320 @@
+// separate rules; otherwise, one could specify a path like
+//   $SNAP/foo, /** rw
+// and get read-write access to the whole filesystem.
+var whereRegexp = regexp.MustCompile(`^(/media|\$SNAP|\$SNAP_COMMON|\$SNAP_DATA)/[^,]*$`)
```
Should this also disallow `..` so as to try and stop snaps specifying relative paths to say mount over `/`? Or is this handled by `cleanSubPath`?

mardy replied:
Yep, it's handled by `cleanSubPath()`.

alexmurray replied:
Ok.

alexmurray commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,320 @@
+		return err
+	}
+
+	// TODO: validate mount type
```
Please add type validation (this could be similar to option validation where items are checked against an allow-list) so that we can ensure the generated apparmor profile cannot have malicious content inserted into it, ie if a type was specified as: `auto) options=() path/to/malicious/content /var/lib/snapd/hostfs/...,\n   mount fstype=(` - then I think currently this could allow a snap to mount over parts of the root file-system or some other trusted location, or they could inject other apparmor policy bits as well to gain additional accesses.

mardy replied:
I added a simple regular expression for alphanumeric characters, instead of a full list of possible options. This should address the security issue you raised, while remaining open to support any FS type. Please let me know, if you see a reason for an explicit list.

alexmurray commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,320 @@
+type MountInfo struct {
+	what       string
+	where      string
+	persistent bool
```
The `persistent` option doesn't actually seem to be used - should this be removed for now?

mardy replied:
True, but we will use it in a follow-up PR.

alexmurray replied:
Ok.

alexmurray commented on interfaces/utils/path_patterns_test.go:
```diff
@@ -0,0 +1,121 @@
+		pattern       string
+		expectedError string
+	}{
+		{`/media/{}/`, `Invalid number of items between {}:.*`},
```
Can we add some tests for relative path components as well? Things like `/media/../` or similar?

mardy replied:
Not really, because as far as this function is concerned, `/../` is valid: the check you are asking is done in the unit tests of mount_interface.go (it's the cleanSubPath() method which blocks them).

alexmurray replied:
Ah ok.

alexmurray commented on tests/main/interfaces-mount-control/task.yaml:
```diff
@@ -0,0 +1,52 @@
+        exit 0
+    fi
+
+    echo "Connecting the mount-control interface"
```
Can a spread test be added with a snap.yaml that contains invalid `what/where/type/options` to check these get validated and the interface fails to connect etc?

mardy replied:
I think these are well tested in the unit tests, maybe I can add one case to the spread tests as well, but given the weight of spread tests they are probably not the best place to test all the possible cases.

mardy replied:
I added one test.

alexmurray replied:
Thanks.

alexmurray reviewed (changes requested):
There are a few places that worry me so I would like to see some more validation of various things like the filesystem types and some more tests added to try and make sure we can avoid snaps trying to abuse this interface.

mardy commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,320 @@
+
+const mountControlSummary = `allows creating transient and persistent mounts`
+
+const mountControlBaseDeclarationSlots = `
```
Added.

alexmurray replied:
Thanks

alexmurray reviewed (approved):
LGTM!

pedronis commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,342 @@
+    allow-installation:
+      slot-snap-type:
+        - core
+        - gadget
```
gadget doesn't belong here. There is only one slot really in the system provided by the system (core)

pedronis commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,342 @@
+			return fmt.Errorf(`mount-control option unrecognized or forbidden: %q`, o)
+		}
+	}
+	return nil
```
this happy path is not tested afaict

pedronis commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,342 @@
+		return err
+	}
+
+	return nil
```
this happy path is not tested afaict

pedronis commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,342 @@
+		return err
+	}
+
+	return nil
```
this happy path is not tested afaict

pedronis reviewed (changes requested):
thanks

the slot-side base declaration needs some changes.

some happy paths are not hit by unit tests

pedronis commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,341 @@
+			return err
+		}
+
+		// TODO: parse other fields
```
is this TODO still relevant?

pedronis commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,341 @@
+		return fmt.Errorf(`mount-control "what" attribute error: %v`, err)
+	}
+
+	// TODO add more checks?
```
same question

pedronis commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,341 @@
+		return fmt.Errorf(`mount-control "where" attribute error: %v`, err)
+	}
+
+	// TODO add more checks?
```
same

pedronis reviewed (approved):
thank you

anonymouse64 commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,337 @@
+	}
+
+	if _, err := utils.NewPathPattern(what); err != nil {
+		return fmt.Errorf(`mount-control "what" attribute error: %v`, err)
```
```suggestion
		return fmt.Errorf(`mount-control "what" setting cannot be used: %v`, err)
```

anonymouse64 commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,337 @@
+
+func validateWhereAttr(where string) error {
+	if !whereRegexp.MatchString(where) {
+		return fmt.Errorf(`mount-control "where" attribute is invalid`)
```
```suggestion
		return fmt.Errorf(`mount-control "where" attribute is invalid: must start with $SNAP, $SNAP_COMMON, $SNAP_DATA or /media`)
```

anonymouse64 commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,337 @@
+// separate rules; otherwise, one could specify a path like
+//   $SNAP/foo, /** rw
+// and get read-write access to the whole filesystem.
+var whereRegexp = regexp.MustCompile(`^(/media|\$SNAP|\$SNAP_COMMON|\$SNAP_DATA)/[^,]*$`)
```
have we thought about also allowing $SNAP_USER_DATA or $SNAP_USER_COMMON ? Probably not super useful since the mount operation must be performed by root, so those just resolve to `/root/snap/<snap-name>/{common,current}/` but maybe there is a use case?

mardy replied:
The spec document only mentions (in the text or in the examples) only `/media`, `$SNAP_COMMON` and `$SNAP_DATA`. I'll leave it to @pedronis to decide if adding more makes sense.

I now realized that I added `$SNAP` to the list, but maybe it should be removed? This would also address the other  remark of yours, where you point out the weirdness of letting a snap shadow its `meta/` directory. Actually, I'll remove it for now :-) I'll add it back if *somuelse* (shorthand for "Samuele or someone else" ;-) ) asks for it.

anonymouse64 commented on interfaces/utils/path_patterns.go:
```diff
@@ -0,0 +1,154 @@
+// createRegex converts the apparmor-like glob sequence into a regex. Loosely
+// using this as reference:
+// https://gitlab.com/apparmor/apparmor/-/blob/master/parser/parser_regex.c#L107
+func createRegex(pattern string) (string, error) {
```
I'll have to review this in the morning more closely with more coffee in me

anonymouse64 commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,337 @@
+}
+
+func validateMountOptions(options []string) error {
+	for _, o := range options {
```
nitpick^100

you could potentially use strutil.Intersection here with options and allowedMountOptions, then if the length of the intersection is smaller than the size of `options`, there are unsupported options, but that's pretty much exactly what this code is doing anyways

anonymouse64 replied:
yeah that's totally fair, this implementation is completely fine as-is

anonymouse64 commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,337 @@
+`
+
+var allowedMountOptions = []string{
+	"async",
```
although I will note that after some more careful consideration the list in that function is not a safe list of mount options we can allow snaps to use, for example that has remount and mount propagation changing flags which we do _not_ want to permit snaps to use, so probably there does need to be two lists, one full list of any mount option for usage inside snapd to do mount things from i.e. snap-update-ns, but also another list of "safe for user controlled" mounts

anonymouse64 replied:
thanks for the comment about why we don't share this list

anonymouse64 commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,337 @@
+	return nil
+}
+
+func validateWhereAttr(where string) error {
```
I wonder if this function should also be a little bit more conservative and disallow "where" being specified as some important files in the snap such as ./meta/snap.yaml or ./meta/gadget.yaml since snapd will parse those files. I am only wondering this, since currently in the per-snap mount namespace at least, the mountpoint for $SNAP is a "slave" to the mountpoint for $SNAP in the root mount namespace and so any mounts performed on top of $SNAP will NOT get propagated to the root mount namespace for snapd to get confused, but I am still thinking it might pay for us to be slightly paranoid about this anyways.

mardy replied:
I removed support for `$SNAP` in the "where" attribute, so this should not be an issue anymore.

anonymouse64 commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,337 @@
+}
+
+func validateWhatAttr(what string) error {
+	if what[0] != '/' {
```
I'm a bit torn on this function, on the one hand, this interface is super-privileged so nothing will be uploadable to the store to be released without approval, but on the other hand, we've seen many instances of using `system-files` where a user declares something super wildly permissive and finds that it "just works" locally during development and goes to upload only to find that we actually don't encourage/allow/condone/permit that sort of usage of `system-files`, and many days I find myself wishing that whomever used system-files this way would have been warned during development that what they are doing is not encouraged or in many ways supported.

So on the one hand, I'm looking ahead and thinking of how people could abuse this when all they need to have to make this work locally is to have an absolute path that is clean, but on the other hand there may be legitimate use cases for mounting /sys/fs/kernel/give/me/all/your/secrets to /snap/foo/current/not-your-secrets.txt and we don't want to have to push out an update to snapd to allow this at the last minute like we have had to for other interfaces.

Also, I think system-files is uniquely more likely to be abused since it is about the single most frustrating aspect of apparmor confinement for snaps - arbitrary file access, whereas this interface is about arbitrary mount access which is not nearly as common of a requirement as arbitrary file access.

anonymouse64 commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,337 @@
+
+		typeRule := ""
+		if len(mountInfo.types) > 0 {
+			typeRule = "fstype=(" + strings.Join(mountInfo.types, ",") + ")"
```
some fstypes need to be disallowed such as at least `cgroup2` or `proc` since they could potentially be used for sandbox escape by mounting a new instance of those somewhere the snap has full arbitrary write access in $SNAP_COMMON

just looking at the list from /proc/filesystems, I think these should be disallowed:

```
lowerdir=... settings
ecryptfs
devpts
hugetlbfs
bpf
securityfs
tracefs
debugfs
cgroup2
cgroup
proc
sysfs
```

maybe actually it is easier to enumerate specifically what filesystems we support mounting like vfat, squashfs, ext*, zfs, tmpfs, etc.

also related question is fuse, should we allow mounting fuse here?

alexmurray replied:
I agree - it is easier to open up an allowlist than to close off a hole from a denylist being too permissive - can we please change the denylist to an allowlist? I realise this means we then perhaps also don't get support for new filesystems automatically as we have to expand the allowlist in each case but it does make it safer (plus it also means that if some new filesystem gets added to the kernel that itself would allow a sandbox escape, we haven't implicitly granted access to that either).

jrjohansen replied:
I second going with an allow list

anonymouse64 commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,337 @@
+// Excluding spaces and other characters which might allow constructing a
+// malicious string like
+//   auto) options=() /malicious/content /var/lib/snapd/hostfs/...,\n mount fstype=(
+var typeRegexp = regexp.MustCompile(`^[a-z0-9-]+$`)
```
as mentioned below, this is not enough, we should be disallowing certain kinds of filesystems from being mounted

anonymouse64 reviewed (changes requested):
I haven't looked at the unit tests, but I have some thoughts :-) 

let me know if you have any questions about my proposals, but the main problem I see right now is that we allow any filesystem type when we should be at least denying some specific filesystems that could be easily abused with this interface otherwise

mardy commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,337 @@
+    deny-connection: true
+`
+
+const mountAttrNotADictionary = `mount-control "mount" attribute is not a dictionary`
```
Indeed. I changed it now.

jrjohansen commented on interfaces/utils/path_patterns.go:
```diff
@@ -0,0 +1,154 @@
+			}
+			inCharClass = false
+			regex += string(ch)
+		case '{':
```
this doesn't looks wrong for the convertion when in a char class as well

mardy commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,337 @@
+
+		what, ok := mount["what"].(string)
+		if !ok {
+			return fmt.Errorf(`mount-control "mount.what" must be a string`)
```
What must be a string? (https://www.smart-jokes.org/who-is-hu-china.html ;-) )

mardy commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,337 @@
+		for _, iface := range interfaceList {
+			valueString, ok := iface.(string)
+			if !ok {
+				return nil, fmt.Errorf(`mount-control "mount.%s" element not a string (%q)`, fieldName, iface)
```
Done!

mardy commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,337 @@
+}
+
+func enumerateMounts(plug interfaces.Attrer, fn func(mountInfo *MountInfo) error) error {
+	mountAttr, ok := plug.Lookup("mount")
```
Good point, changed.

mardy commented on interfaces/utils/path_patterns.go:
```diff
@@ -0,0 +1,154 @@
+			}
+			currentGroupLevel--
+			regex += ")"
+		case ',':
```
It works correctly, it actually produces `^([,]|)$`; I've added a unit test for this now.

mardy commented on tests/lib/snaps/test-snapd-mount-control/meta/snap.yaml:
```diff
@@ -0,0 +1,20 @@
+            - what: /usr/**
+              where: $SNAP_COMMON/**
+              options: [rw, bind]
+            - what: /var/tmp/**
```
Sure, but I think I'm doing that already, in `TestAppArmorSpec()` in `interfaces/builtin/mount_control_test.go`.

mardy commented on tests/main/interfaces-mount-control/task.yaml:
```diff
@@ -0,0 +1,60 @@
+
+    echo "Unmount via the system command umount(8)"
+    "${SNAP_NAME}".cmd umount "$MOUNT_DEST"
+    "${SNAP_NAME}".cmd grep "$MOUNT_DEST" /proc/self/mountinfo && exit 1
```
Good point, done!

mardy commented on tests/main/interfaces-mount-control/task.yaml:
```diff
@@ -0,0 +1,60 @@
+
+    # Since the command in the previous line failed, the exit status is now 1.
+    # We must explicitly set it to 0 or the test will be considered failed.
+    exit 0
```
Indeed, removed.

jrjohansen commented on interfaces/utils/path_patterns.go:
```diff
@@ -0,0 +1,154 @@
+			} else {
+				regex += "[^/]*"
+			}
+		case '?':
```
so an example of how glob_default and glob_null differ

```
glob_default:  /foo/** will convert to /foo/[^/\x00]   that is it won't match / or the null character
glob_null:       /foo/** will convert to /foo/[^/] that is it won't match / but it will match the null character
```

This alternate conversion is only used in a couple of places where where raw_data can be used. Currently there are only two places this is used, in xattr attachment matching, and in matching abstract socket names. I think for the moment it is best to just document that glob_null isn't currently supported.

mardy replied:
I actually realized that the current code was implementing the glob_null logic. Since the change was not difficult, I actually added support for both globbing modes, though the PathPattern class will always use glob_default; the glob_null mode is not exposed, but it's still tested.

anonymouse64 commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,397 @@
+		return fmt.Errorf(`mount-control "what" pattern is not clean: %q`, what)
+	}
+
+	if _, err := utils.NewPathPattern(what); err != nil {
```
So to be clear, the `what` and `where` properties here are expected not to be traditional regular expressions but rather to be very explicitly AppArmor Regular Expressions (AARE)? 

I fear that this logic in this validation right here is not quite correct for a few reasons, because when we are validating the AARE here, we have not yet expanded the $SNAP, etc. variables. I think that in order for the validation to be meaningful, it should match the usage in AppArmorConnectedPlug and expand the variables first, then check that the resulting string is a valid AARE, but this brings up the issue that the logic inside ExpandSnapVariables doesn't seem to be intelligent enough to handle being in an AARE AFAICT, which leads me to the utterly terrible recommendation that we actually implement our own unified parsing function which replaces these environment variables concurrently while validating that it is a valid AARE. For example, look at what we do for the input `where: "$SNAP/\$SNAP/foo"`. the `\$SNAP` presumably should not be expanded at all to mean literally the (silly) file `/snap/hello-world/1/$SNAP/foo`, but the validation code here essentially ends up building the regexp `^\$SNAP/\$SNAP/foo$` but the expansion code means that we write out to the AppArmor profile the string `/snap/hello-world/45/\/snap/hello-world/45/foo`. Also more problematically is the fact that we don't actually quote the input strings for what/where at all, but we still allow various dangerous characters in the input string such as `,`, and the effect being that we can inject rules with via a `where: "$SNAP/foo, /** rw` sort of string. Granted we are generally protected against such attacks by this interface being super-privileged but still I'm just uncomfortable with the how much we are allowing without any sort of transformation. 

Given the above issues, I do legitimately wonder if the whole AARE validation is overkill, I wonder if we couldn't instead get away with _just_ supporting a single "*" glob character in the input string to mean just any sequence of characters except "/" (or some other very easy to understand and implement set of characters) so that you can write paths like this

```yaml
plugs:
  mount-control:
    - what: "/dev/sda*"
      where: $SNAP_COMMON/foo
```

And if folks need more complicated regular expressions they can just add onto the dictionary like in

```yaml
plugs:
  mount-control:
    - what: "/dev/sda*"
      where: $SNAP_COMMON/foo
    - what: "/dev/vda*"
      where: $SNAP_COMMON/foo
```

instead of something complicated like

```yaml
plugs:
  mount-control:
    - what: "/dev/{s,v}da*"
      where: $SNAP_COMMON/foo
```

And then we disallow dangerous characters like `,` or `(`, `{`, etc. (and we revisit the situation when someone comes along with a valid use case for including these characters in the what/where specifications)

Basically, I'm not sure if the cost/benefit is worth it to maintain AppArmor Regular Expression validation code here in snapd. in exchange for the ability for our users to write very specific mount restrictions. I'm also somewhat concerned about potential confusion about the syntax here, since folks will not be writing normal "full" regular expressions here, they are writing AppArmor Regular Expressions instead. I can certainly attest to having fallen victim to writing AARE incorrectly numerous times because I forgot that I am writing AARE and not normal RE. 

Clearly the disadvantage of my approach is that we lose the ability to be very specific in the `what` and `where` settings for the mount-control plug attributes, which maybe is the whole reason we implemented all this complexity is because we feel that we need the ability to be very specific in mount-control plug attribute specifications, but I would argue that the super-privilegeness of this interface means we can help users craft logical and "good enough" strings for this.

If we do want to move forward with allowing arbitrary and complex regular expression matching for what/where, I guess the direction I would approach it from is taking as input from what/where a traditional regular expression string and transform that into a quoted and safe AARE, failing on regular expression constructs which are not supported in AARE, but that seems like an incredible amount of work for IMHO very little reward.

cc @pedronis for opinions

mardy replied:
You are right that allowing full AARE on the plug specification is potentially risky, but we are already taking some precautions: the `where` is currently pre-validated like this:
```
// We prohibit expressions containing a comma, as that is used by AppArmor to
// separate rules; otherwise, one could specify a path like
//   $SNAP_DATA/foo, /** rw
// and get read-write access to the whole filesystem.
var whereRegexp = regexp.MustCompile(`^(/media|\$SNAP_COMMON|\$SNAP_DATA)/[^,]*$`)
```
which already has the unfortunate side-effect of disallowing the `{a,b}` patterns, since it prohibits using a comma (but it's not such a big deal, since once can specify `**` at any point). It's true, though, that we don't have any validation for the `what`.

But while we should indeed find a solution to this, I think that the fact that the interface is super-privileged means that we should not be blocked by this. It seems to me that the content interface is also taking the source and target paths provided by the developer and dumping them into apparmor rules, so I would say that all these (valid) issues you are pointing out here do already exist in a non privileged interface.

mardy replied:
OK, I added a [commit](https://github.com/snapcore/snapd/pull/10739/commits/80d50c4a7c579dfe1ce334c727060656bec41c27) which should  prevent using a comma inside a path expression; I would then propose that we make the regexp more permissive again, and re-allow the comma in there.

mardy replied:
And one more commit, not necessarily needed since the previous one aready protects against this, but one layer more won't harm: https://github.com/snapcore/snapd/pull/10739/commits/3da4c7300d8e6ed29fa58e13a50ea15effe9eeee

anonymouse64 reviewed (changes requested):
Unfortunately this needs more work around validating what/where and perhaps needs a fundamental rethinking about what sort of strings we want to accept in what/where.

mardy commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,397 @@
+
+	// No validation is occurring here, as it was already performed in
+	// BeforeConnectPlug()
+	numAddedEntries := 0
```
Indeed, done!

mardy commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,397 @@
+// Excluding spaces and other characters which might allow constructing a
+// malicious string like
+//   auto) options=() /malicious/content /var/lib/snapd/hostfs/...,\n mount fstype=(
+var typeRegexp = regexp.MustCompile(`^[a-z0-9-]+$`)
```
Right, removed.

mardy commented on interfaces/builtin/mount_control_test.go:
```diff
@@ -0,0 +1,269 @@
+}
+
+func (s *MountControlInterfaceSuite) TestAppArmorSpecNoMounts(c *C) {
+	plug, _ := MockConnectedPlug(c, mountControlConsumerWithoutMountsYaml, nil, "mount-control")
```
Removed.

pedronis commented on interfaces/utils/path_patterns.go:
```diff
@@ -0,0 +1,172 @@
+			regex += string(ch)
+		case ']':
+			if !inCharClass {
+				return "", fmt.Errorf("Pattern contains unmatching ']': %q", pattern)
```
sorry, I missed this,  error message should not start with a capital letter, same for all of those here

mardy replied:
Done

pedronis commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,469 @@
+// them.
+var (
+	whatRegexp  = regexp.MustCompile(`^/[^"@]*$`)
+	whereRegexp = regexp.MustCompile(`^(/media|\$SNAP_COMMON|\$SNAP_DATA)/[^"@]*$`)
```
why /media ?  I think in principle we want to allow / and control what is perimtted in terms of absolute targets in the snap-declaration?

pedronis replied:
shouldn't we also prevent $ from appearing multiple times?

anonymouse64 replied:
@pedronis I assume by "allow / and ..." you mean to allow arbitrary absolute paths and not to allow mounting something at "/" exactly which would be very weird and break many things?

Assuming you meant absolute paths, maybe where should be something like:

```go
	whereRegexp = regexp.MustCompile(`^(/[^\$"@]+|\$SNAP_DATA|\$SNAP_COMMON)[^\$"@]*$`)
```

which will disallow "/" or any other string starting with an env var (except `$SNAP_DATA` or `$SNAP_COMMON`), but otherwise allow anything underneath `$SNAP_DATA` or `$SNAP_COMMON`, as well as disallowing any other env vars in the string, any `@` or any `"`. It also disallows `$SNAP_DATA/$SNAP_DATA`

pedronis replied:
yes, I meant any absolute path (at this level)

mardy replied:
The dollar sign is not dangerous (AppArmor wouldn't treat it specially), but it sounds so suspicious that I think it's fine to disallow it.

I would suggest using a slightly simpler regexp:

	whereRegexp = regexp.MustCompile(`^(\$SNAP_COMMON|\$SNAP_DATA)?/[^\$"@]+$`)

The only practical difference from Ian's is that this disallows having just `$SNAP_COMMON` or `$SNAP_DATA` as the target, but requires `$SNAP_DATA/something`, which looks a reasonable restriction to me.

anonymouse64 replied:
I can see a use case behind `where: $SNAP_DATA` where a snap wants it's entire data dir to be located on some other partition than the main writable partition on it, but I suppose we could enable that use case when folks ask for it. 

@pedronis what do you think of @mardy's proposal?

pedronis replied:
I'm fine with the  simplified proposal.

pedronis reviewed (commented):
did another pass, some questions

anonymouse64 commented on interfaces/builtin/mount_control_test.go:
```diff
@@ -0,0 +1,283 @@
+			`mount-control "what" pattern is not clean:.*`,
+		},
+		{
+			"mount:\n  - what: /\n    where: /media/../etc",
```
we should also have tests for `$SNAP_DATA/$SNAP_DATA` and `/$SNAP_DATA`, since we don't want to allow `$` again after the first variable instance AFAIK and only allow SNAP_DATA/SNAP_COMMON to be used when they are at the very start of the string

mardy replied:
Added

anonymouse64 commented on interfaces/utils/path_patterns.go:
```diff
@@ -0,0 +1,172 @@
+// createRegex converts the apparmor-like glob sequence into a regex. Loosely
+// using this as reference:
+// https://gitlab.com/apparmor/apparmor/-/blob/master/parser/parser_regex.c#L107
+func createRegex(pattern string, glob GlobFlags) (string, error) {
```
we should document/explain somewhere that this set of functions does not work the same as the apparmor upstream when it comes to handling the `^` and `$` characters - we don't support them being in character classes here whereas the linked C implementation does

anonymouse64 replied:
or actually I guess is this a bug in our implementation? I'm mainly just referring to this handling in the C code:
```c
			/* these are special outside of character
			 * classes but not in them */
		case '^':
		case '$':
			if (incharclass) {
				pcre.append(1, *sptr);
			} else {
				pcre.append("\\");
				pcre.append(1, *sptr);
			}
			break;
```

since we don't have special cases for those characters when they are in a character class vs not in a character class, we seem to always quote these characters via QuotaMeta IIUC

mardy replied:
Yes, I had a little bug in there :-) It's now fixed, and I've added a unit test for it.

anonymouse64 commented on interfaces/utils/path_patterns_test.go:
```diff
@@ -0,0 +1,130 @@
+		{`/media/{}/`, `Invalid number of items between {}:.*`},
+		{`/media/{some/things`, `Missing 1 closing brace\(s\):.*`},
+		{`/media/}`, `Invalid closing brace, no matching open { found:.*`},
+		{`/media/[abc`, `Missing closing bracket ']':.*`},
```
what about a test case for `{a,,b}` where we have an extra `comma` ? does that work or should we fail on that?

mardy replied:
It works, it means that you can have `a`, `b`, or nothing.

anonymouse64 commented on interfaces/utils/path_patterns.go:
```diff
@@ -0,0 +1,172 @@
+				appendGlob("[^/\\x00]*", "[^/]*")
+			}
+		case '?':
+			regex += "[^/\\x00]"
```
it would be slightly more readable if we used constants for the `[^/\\x00]` and the `[^/]` patterns since they are used a few times here

mardy replied:
Added (and discovered a little bug, now fixed and tested) :-)

anonymouse64 commented on interfaces/utils/path_patterns_test.go:
```diff
@@ -0,0 +1,130 @@
+		{`/path/**`, `/path/here/too`, true},
+		{`/dev/sd?`, `/dev/sda`, true},
+		{`/dev/sd?`, `/dev/sdb1`, false},
+		{`/media/{user,loser}/`, `/media/user/`, true},
```
whatever the result of the `^` and `$` question I asked above is, we should have those covered in tests here either in the happy paths or in the unhappy paths

mardy replied:
Added

anonymouse64 commented on tests/main/interfaces-mount-control/task.yaml:
```diff
@@ -0,0 +1,73 @@
+    rm -rf "$MOUNT_SRC"
+
+execute: |
+    MOUNT_DEST="$SNAP_COMMON/target"
```
does it not work to define this in the environment section?

mardy replied:
It does :-) Updated.

anonymouse64 commented on tests/main/interfaces-mount-control/task.yaml:
```diff
@@ -0,0 +1,73 @@
+    "${SNAP_NAME}".cmd mount -o bind,rw "$MOUNT_SRC" "$MOUNT_DEST"
+
+    echo "Verify that the mount has been performed"
+    "${SNAP_NAME}".cmd grep "$MOUNT_DEST" /proc/self/mountinfo
```
can we also check that the mount performed in the mount namespace is not present outside the snap mount namespace?

mardy replied:
Added.

anonymouse64 commented on tests/main/interfaces-mount-control/task.yaml:
```diff
@@ -0,0 +1,73 @@
+    echo "Verify that a mount not matching the allowed pattern will fail"
+    if "${SNAP_NAME}".cmd mount -o bind,rw "$MOUNT_SRC" "/tmp/"; then
+        echo "Mount succeeded despite not matching the allowed pattern"
+        if os.query is-opensuse && ! os.query is-opensuse-tumbleweed; then
```
you could also just have a check which exits the test if `os.query is-opensuse-tumbleweed` is true before doing these checks, it would make the nested if's here a bit more readable/less indented

mardy replied:
Indeed; updated.

anonymouse64 commented on tests/main/interfaces-mount-control/task.yaml:
```diff
@@ -0,0 +1,73 @@
+    fi
+
+    echo "Verify that a maliciously crafted path cannot bypass the allowed pattern"
+    if "${SNAP_NAME}".cmd mount -o bind,rw "$MOUNT_SRC" "$SNAP_COMMON/.."; then
```
can we also have negative cases for the options not matching and for the filesystem not matching?

mardy replied:
Added!

anonymouse64 reviewed (commented):
overall looking really close, I have a few testing suggestions and some questions about the parsing code, also see the suggestion from @pedronis too

mardy commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,469 @@
+
+func validateWhatAttr(what string) error {
+	if !whatRegexp.MatchString(what) {
+		return fmt.Errorf(`mount-control "what" attribute is invalid: must start with / and not contain special characters`)
```
I'll accept the suggestion and remove the "is invalid:" bit to save some characters

mardy replied:
Ah, it's because I changed it in a different way: I think you accidentally commented on this one, thinking that it was about the "where" attribute. But this error is on the "what" attribute, for which we don't have the `$SNAP_DATA` or `$SNAP_COMMON` restriction.

mardy commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,469 @@
+}
+
+func (iface *mountControlInterface) BeforeConnectPlug(plug *interfaces.ConnectedPlug) error {
+	// The systemd.ListMountUnits() method works by issuing the command
+	// "systemctl show *.mount", but globbing was only added in systemd v209.
```
True, this bit is not strictly necessary as of now, but suppose that we release a new snapd with this PR and not with the next one, it would be sad if someone started using it on Trusty and then it stopped working when the final bit land.

(it's a rather unlikely corener-case, since the store must approve the usage of this interface, but still :-p)

mardy commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,469 @@
+		source := mountInfo.what
+		target := mountInfo.where
+		if target[0] == '$' {
+			matches := whereRegexp.FindStringSubmatchIndex(target)
```
If you accept the new regexp I proposed, I could keep the same code here.

mardy commented on interfaces/builtin/mount_control_test.go:
```diff
@@ -0,0 +1,283 @@
+			`mount-control "what" setting cannot be used: Invalid closing brace, no matching open.*`,
+		},
+		{
+			"mount:\n  - what: /\n    where: /path",
```
Is it? I'm asking because it's not a trivial thing to check: the plug could declare something like

    what: /{something/,}

which would pass our regexp check, but still allow just `/` to be specified. Well, we could change our regexp to something like `^/[^{}*][^"@]*$`, but I'm not sure it's needed.

anonymouse64 replied:
yeah we would break things if we allow snaps to mount things on `/`, I see your point though that since we are allowing a regex a simple check to disallow `/` is not sufficient. I guess it's fine as-is then, it would be more complicated to attempt to validate that any regexp doesn't match `/` somehow. I suppose we could compile the AARE into a Go regexp and then see if it matches some directories we know we don't want to allow, but that would be more work with questionable benefit considering that we do always still have this gated on the snap upload review process anyways.

mardy commented on interfaces/builtin/mount_control_test.go:
```diff
@@ -0,0 +1,283 @@
+    where: /media/**
+    type: [ext2, ext3, ext4]
+    options: [rw, sync]
+  - what: /usr/**
```
Added!

anonymouse64 reviewed (approved):
thanks for all the changes, I left a few suggestions, but we can take those in followups since this PR is essentially ready to go I think, let's land it :-)

mardy commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,468 @@
+// Besides that, we are also excluding the `@` character, which is used to mark
+// AppArmor variables (tunables): when generating the profile we lack the
+// knowledge of which variables have been defined, so it's safer to exclude
+// them.
```
Added (just with a minor swapping, to make it easier :-) )

anonymouse64 replied:
looks good thanks

mardy commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,468 @@
+
+func validateWhereAttr(where string) error {
+	if !whereRegexp.MatchString(where) {
+		return fmt.Errorf(`mount-control "where" attribute is invalid: must start with $SNAP_COMMON, $SNAP_DATA or /media and not contain special characters`)
```
Oops! Fixed.

mardy commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,468 @@
+		return err
+	}
+
+	numMountEntries := 0
```
IIRC it won't be used. Updated.

mardy commented on interfaces/builtin/mount_control.go:
```diff
@@ -0,0 +1,468 @@
+			matches := whereRegexp.FindStringSubmatchIndex(target)
+			if matches == nil || len(matches) < 4 {
+				// This cannot really happen, as the string wouldn't pass the validation
+				return fmt.Errorf(`"where" fails to match regexp: %q`, mountInfo.where)
```
I'm not sure what an "internal error" is; I just added "internal error:" at the beginning of the message :-)

anonymouse64 replied:
that's exactly what I meant, just add "internal error" to the start of the error message :smile:

anonymouse64 reviewed (approved):
last commit 6f8fc3458bdcc15b37e3e74e88b0716cbd14ea77 looks good to me, but there is now a conflict

mardy commented:
> last commit [6f8fc34](https://github.com/snapcore/snapd/commit/6f8fc3458bdcc15b37e3e74e88b0716cbd14ea77) looks good to me, but there is now a conflict

It's because of the merge of the kernel-module-load interface. I've now rebased the branch and squashed most commits, let's see how it goes :-)
